Small improvements to the Numbers page (#4649)
JetBrains/kotlin-web-site: docs/topics/numbers.md

## Integer types

Kotlin provides a set of built-in types that represent numbers.  
For integer numbers, there are four types with different sizes and value ranges:

| Type	    | Size (bits) | Min value                                    | Max value                                      |
|----------|-------------|----------------------------------------------|------------------------------------------------|
| `Byte`	  | 8           | -128                                         | 127                                            |
| `Short`	 | 16          | -32768                                       | 32767                                          |
| `Int`	   | 32          | -2,147,483,648 (-2<sup>31</sup>)             | 2,147,483,647 (2<sup>31</sup> - 1)             |
| `Long`	  | 64          | -9,223,372,036,854,775,808 (-2<sup>63</sup>) | 9,223,372,036,854,775,807 (2<sup>63</sup> - 1) |

> In addition to signed integer types, Kotlin also provides unsigned integer types.
> As unsigned integers are aimed at a different set of use cases, they are covered separately.

> > See the [](unsigned-integer-types.md) page.

Reviewer: ```suggestion
> See [](unsigned-integer-types.md).
```
Nitpick.

Kotlin also supports a conventional notation for floating-point numbers:

* Doubles (fractional part without other tags): `123.5`, `123.5e10`

> * Floats, tagged by `f` or `F`: `123.5f`

Author: I didn't think to check, but I don't think there's an established word. The official docs only says "floating-point literal is of type `float` if it ends with the letter F or f", and everyone else kind of does the same.

I would probably choose "suffix" as something to call these letters, although in this particular case "...ending with `f` or `F`" sounds more natural to me.

> * Floats, tagged by `f` or `F`: `123.5f`

Reviewer: Using "suffix" or "ending with" works for me :)

> * Floats, tagged by `f` or `F`: `123.5f`

Author: Most of the "tag" usage is in the unsigned numbers article, so I adjusted the wording there as well, check it out.

val socialSecurityNumber = 999_99_9999L
val hexBytes = 0xFF_EC_DE_5E
val bytes = 0b11010010_01101001_10010100_10010010
val bigFractional = 1_234_567.7182818284
```

> There are also special tags for unsigned integer literals.  
> Read more about [literals for unsigned integer types](unsigned-integer-types.md).
> 
{style="tip"}

## Boxing numbers on the Java Virtual Machine

JVM stores numbers as primitive types: `int`, `double`, and so on.
When you use generics or create a nullable number reference such as `Int?`, numbers are boxed in Java classes
such as `Integer` or `Double`.

JVM applies a [memory optimization](https://docs.oracle.com/javase/specs/jls/se22/html/jls-5.html#jls-5.1.7)
to `Integer` objects that represent numbers between `−128` and `127`:

> all nullable references to such objects are referring to the same object.

Author: Yes, the basic point is that JVM caches objects for primitive types that fit in a byte. So when we create a new variable without explicitly requesting a separate object, JVM checks if there's already a cached object with that value and returns the reference to it if there is.

I tried to get deeper into the subject, not sure if that made it clearer.

val socialSecurityNumber = 999_99_9999L
val hexBytes = 0xFF_EC_DE_5E
val bytes = 0b11010010_01101001_10010100_10010010
val bigFractional = 1_234_567.7182818284
```

> > There are also special tags for unsigned integer literals.  

Reviewer: Can we remove the mention of "tags" here as well?

> Read more about [literals for unsigned integer types](unsigned-integer-types.md).
> 
{style="tip"}

## Boxing and caching numbers on the Java Virtual Machine

> The way JVM stores numbers can make your code behave counterintuitively because of the cache used by default

Reviewer: ```suggestion
The way the JVM stores numbers can make your code behave counterintuitively because of the cache used by default
```

for small (byte-sized) numbers.

> JVM stores numbers as primitive types: `int`, `double`, and so on.

Reviewer: ```suggestion
The JVM stores numbers as primitive types: `int`, `double`, and so on.
```

When you use [generic types](generics.md) or create a nullable number reference such as `Int?`, numbers are boxed in Java classes
such as `Integer` or `Double`.

> JVM applies a [memory optimization](https://docs.oracle.com/javase/specs/jls/se22/html/jls-5.html#jls-5.1.7)

Reviewer: ```suggestion
The JVM applies [memory optimization](https://docs.oracle.com/javase/specs/jls/se22/html/jls-5.html#jls-5.1.7)
```

> to `Integer` and other objects that represent numbers between `−128` and `127`:
> all nullable references to such objects are referring to the same cached object.

Reviewer: ```suggestion
to `Integer` and other objects that represent numbers between `−128` and `127`.
All nullable references to such objects refer to the same cached object.
```
This is much clearer now! But perhaps we could split it into two sentences?

```kotlin
fun main() {
//sampleStart
    val byte: Byte = 1 // OK, literals are checked statically
    // val intAssignedByte: Int = byte // Error: Type mismatch
    val intConvertedByte: Int = byte.toInt()
    
    println(intConvertedByte)
//sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

All number types support conversions to other types:

* `toByte(): Byte` (deprecated for [Float](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/-float/to-byte.html) and [Double](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/-double/to-byte.html))
* `toShort(): Short`
* `toInt(): Int`
* `toLong(): Long`
* `toFloat(): Float`
* `toDouble(): Double`

In many cases, there is no need for explicit conversions because the type is inferred from the context,
and arithmetical operators are overloaded for appropriate conversions, for example:

```kotlin
fun main() {
//sampleStart
    val l = 1L + 3 // Long + Int => Long
    println(l is Long) // true
//sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.5"}

### Reasoning against implicit conversions

> If numbers of different types were converted implicitly, we could sometimes lose equality and identity silently.

Author: True. Added this, only rephrased to make Kotlin the subject here :)

    val boxedA: Int? = a
    val anotherBoxedA: Int? = a
    
    println(boxedA === anotherBoxedA) // true
//sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

> For a number outside of this range the nullable objects are different, but [structurally equal](equality.md#structural-equality):

Reviewer: ```suggestion
For numbers outside this range, the nullable objects are different but [structurally equal](equality.md#structural-equality):
```

```kotlin
fun main() {
//sampleStart
    val b: Int = 10000
    val boxedB: Int? = b
    val anotherBoxedB: Int? = b
    
    println(boxedB === anotherBoxedB) // false
    println(boxedB == anotherBoxedB) // true
//sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

For this reason, Kotlin warns about using referential equality with boxable numbers and literals

> with the following message: "Identity equality for arguments of types ... and ... is prohibited."

Reviewer: ```suggestion
with the following message: `"Identity equality for arguments of types ... and ... is prohibited."`
```
Might be better to make this a code block if it wraps in an ugly way 🤔

> When comparing `Int`, `Short`, `Long`, and `Byte` types (as well as `Char` and `Boolean`), use only
> structural equality check to get consistent results.

Reviewer: ```suggestion
When comparing `Int`, `Short`, `Long`, and `Byte` types (as well as `Char` and `Boolean`), use 
structural equality checks to get consistent results.
```

## Explicit number conversions

Due to different representations, number types _are not subtypes_ of each other.
As a consequence, smaller types are _not_ implicitly converted to bigger types and vice versa.
For example, assigning a value of type `Byte` to an `Int` variable requires an explicit conversion:

```kotlin
fun main() {
//sampleStart
    val byte: Byte = 1
    // OK, literals are checked statically
    
    val intAssignedByte: Int = byte 
    // Initializer type mismatch
    
    val intConvertedByte: Int = byte.toInt()
    
    println(intConvertedByte)
//sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3" validate="false"}

All number types support conversions to other types:

* `toByte(): Byte` (deprecated for [Float](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/-float/to-byte.html) and [Double](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/-double/to-byte.html))
* `toShort(): Short`
* `toInt(): Int`
* `toLong(): Long`
* `toFloat(): Float`
* `toDouble(): Double`

In many cases, there is no need for explicit conversion because the type is inferred from the context,

> and arithmetical operators are overloaded for appropriate conversions. For example:

Reviewer: ```suggestion
and arithmetical operators are overloaded to handle conversions automatically. For example:
```

//sampleStart
    val x = 5L / 2
    println(x == 2L)

>     // But not `x == 2`, as Long (x) cannot be compared to Int (2)

Reviewer: Maybe this line should be added as a separate line `println(x == 2)`? So that the code comment below can be `// true` to match the style of the rest of the doc?
